File: Algorithms/Graph_Algorithms/Kruskal's_MST/python/KruskalMST.py
# Python program for Kruskal's algorithm to find
# Minimum Spanning Tree of a given connected,
# undirected and weighted graph
def takeInput():
    """
    Takes input from stdin
    :return: Graph
    """
    v = int(input("Enter the number of vertices \t"))
    e = int(input("Enter the number of edges \t "))
    edges = []
    for i in range(e):
        print("Enter edge {} in form u v w".format(str(i+1)))
        inputEdge = list(map(int, input().split(" ")))
        edges.append(inputEdge)
    graph = Graph(v,edges)
    return graph


class Graph:
    def __init__(self, vertices,graph ):
        """
        constructor for the graph object
        :param vertices:  No of vertices
        :param graph: dictionary used to store the graph
        # Graph is stored as a list of tuples [(u,v,w)]
        """
        self.V = vertices
        self.graph = graph

    def find(self, parent, i):
        """
        Finds the parent or representative element
        :param parent: Array storing current belongingness of the vertices
        :param i: vertex whose parent is to be found
        :return: parent vertex
        """
        if parent[i] == i:
            return i
        return self.find(parent, parent[i])

    def union(self, parent, rank, x, y):
        """
        union by rank of two sets of x and y
        :param parent: array storing current belongingness of the vertices
        :param rank: array storing current rank of the vertices (used to find the rank of parent)
        :param x: set x
        :param y: set y
        :return:
        """
        xroot = self.find(parent, x)
        yroot = self.find(parent, y)

        # Attach smaller rank tree under root of high rank tree (Union by Rank)
        if rank[xroot] < rank[yroot]:
            parent[xroot] = yroot
        elif rank[xroot] > rank[yroot]:
            parent[yroot] = xroot
        else:
            # If ranks are same, then make one as root
            # and increment its rank by one
            parent[yroot] = xroot
            rank[xroot] += 1


    def KruskalMST(self):
        """
        main function construct MST using Kruskal's Algo
        :return: List of edges in the MST
        """

        result = []  # This will store the resultant MST

        i = 0  # An index variable, used for sorted edges
        e = 0  # An index variable, used for result[]

        # Sorting all the edges in non-decereasing order of the weights
        self.graph = sorted(self.graph, key=lambda item: item[2])

        parent = []
        rank = []

        # Create V subsets with single elements
        for node in range(self.V):
            parent.append(node)
            rank.append(0)

        # Number of edges to be taken is equal to V-1
        while e < self.V - 1:

            # Pick the smallest edge and increment
            # the index for next iteration
            u, v, w = self.graph[i]
            i = i + 1
            x = self.find(parent, u)
            y = self.find(parent, v)

            # If including this edge does't cause cycle,
            # include it in result and increment the index
            # of result for next edge
            if x != y:
                e = e + 1
                result.append([u, v, w])
                self.union(parent, rank, x, y)
            # Else discard the edge
            else:
                continue

        # print the contents of result[] to display the built MST
        print("Edges in the MST are as follows ")
        print("Source \t-> Destination \t| Weight")
        for u, v, weight in result:
            print("%d \t-> %d \t\t|  %d" % (u, v, weight))


graph = takeInput()
graph.KruskalMST()

File: Algorithms/Graph_Algorithms/Kruskal's_MST/KruskalMST.c
// Kruskal's Minimum Spanning Tree
// C Programming

#include<stdio.h>
 
#define MAX 30

// GraphEdgeWeights -> Array storing the weights of edges in the graph.
int GraphEdgeWeights[MAX][MAX];

// nVertices -> No: of vertices in the graph.
int nVertices;
 
// Edge -> Structure for an edge with its source vertex, destination vertex and its weight.
typedef struct Edge{
    int src, dest, weight;
} Edge;

// EdgeList -> Structure for storing an array of Edge along with its size.
typedef struct EdgeList{
    Edge data[MAX];
    int size;
} EdgeList;

EdgeList edgeList;
EdgeList spanList;

// Function name : SortEdges
// Input         : void
// Output        : void
// Logic         : For sorting the edges based on their weights.
void SortEdges(){
    int i, j;
    Edge tempEdge;
    
    for(i=1; i<edgeList.size; i++)
        for(j=0; j<edgeList.size-1; j++)
            if(edgeList.data[j].weight > edgeList.data[j+1].weight){
                tempEdge            = edgeList.data[j];
                edgeList.data[j]    = edgeList.data[j+1];
                edgeList.data[j+1]  = tempEdge;
            }
}

// Function name : FindVertexSet
// Input         : vertexSet array, vertexNo
// Output        : vertexSet
// Logic         : For finding the vertexSet of a node in the spanning tree.
int FindVertexSet(int vertexSet[], int vertexNo){
    return(vertexSet[vertexNo]);
}

// Function name : Union
// Input         : vertexSet array, source node, destination node
// Output        : void
// Logic         : For adding the both nodes to the same vertexSet.
void Union(int vertexSet[], int srcNode, int destNode){
    int i;
    for(i=0; i<nVertices; i++)
        if(vertexSet[i] == destNode)
            vertexSet[i] = srcNode;
}

// Function name : KruskalMST
// Input         : void
// Output        : void
// Logic         : For generating a minimum spanning tree.
void KruskalMST(){

    // Initialising the variables.
    int vertexSet[MAX], i, j, srcNode, destNode;
    
    // Initialising the edgeList based on user input.
    edgeList.size = 0;
    for(i=1; i<nVertices; i++){
        for(j=0; j<i; j++){
            if(GraphEdgeWeights[i][j] != 0){
                edgeList.data[edgeList.size].src     = i;
                edgeList.data[edgeList.size].dest    = j;
                edgeList.data[edgeList.size].weight  = GraphEdgeWeights[i][j];
                edgeList.size++;
            }
        }
    }

    // Sorting the edges based on their weights.
    SortEdges();
    
    // Assigning parents to every node of the tree.
    for(i=0; i<nVertices; i++)
        vertexSet[i] = i;
    
    // Creating the Minimum Spanning Tree.
    spanList.size = 0;
    for(i=0; i<edgeList.size; i++){
        // Find Operation
        srcNode = FindVertexSet(vertexSet, edgeList.data[i].src);
        destNode = FindVertexSet(vertexSet, edgeList.data[i].dest);
        
        if(srcNode != destNode){
            spanList.data[spanList.size] = edgeList.data[i];
            spanList.size++;
            // Union Operation
            Union(vertexSet, srcNode, destNode);
        }
    }
}

// Function name : PrintResult
// Input         : void
// Output        : void
// Logic         : For printing the result that is the cost of traversing the spanning tree.
void PrintResult(){
    int i,cost=0;
    
    for(i=0; i<spanList.size; i++){
        printf("\n%d->%d:%d", 
            spanList.data[i].src, 
            spanList.data[i].dest, 
            spanList.data[i].weight);
        cost = cost + spanList.data[i].weight;
    }
    printf("\n=====================================\n");
    printf("Cost = %d", cost);
}

// Function name : main
// Input         : void
// Output        : void
// Logic         : The driver function of the program.
void main(){
    int i, j, total_cost, weight;
    
    printf("\nNumber of vertices : ");
    scanf("%d",&nVertices);
    
    for(i=0; i<nVertices; i++){
        for(j=0; j<nVertices; j++){
            if(i!=j){
                if(i<j){
                    printf("Edge Weight : ", i, j);
                    scanf("%d", &weight);
                    GraphEdgeWeights[i][j] = weight;
                }
                else{
                    GraphEdgeWeights[i][j] = GraphEdgeWeights[j][i];
                }
            }
            else{
                GraphEdgeWeights[i][j] = 0;
            }
        }

    }
            
    KruskalMST();
    PrintResult();
}
 
